qiskit/pulse/instructions/instruction.py

```diff
@@ -0,0 +1,281 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2019.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""
+``Instruction`` s are single events within a :py:class:`~qiskit.pulse.Schedule`, and can be
+used the same way as :py:class:`~qiskit.pulse.Schedule` s.
+
+Specific instructions should be instantiated rather than a generic ``Instruction``.
+
+For example::
+
+    duration = 10
+    channel = DriveChannel(0)
+    sched = Schedule()
+    sched += Delay(duration, channel)  # Delay is a specific subclass of Instruction
+"""
+import warnings
+
+from typing import Any, Tuple, List, Iterable, Callable, Optional, Union
+
+from qiskit.pulse.channels import Channel
+from qiskit.pulse.interfaces import ScheduleComponent
+from qiskit.pulse.schedule import Schedule
+from qiskit.pulse.timeslots import Interval, Timeslot, TimeslotCollection
+
+# pylint: disable=missing-return-doc
+
+
+class Instruction(ScheduleComponent):
+    """The smallest schedulable unit: a single instruction. It has a fixed duration and specified
+    channels.
+    """
+
+    def __init__(self, duration: Union['Command', int],
+                 *channels: Channel,
+                 name: Optional[str] = None):
+        """Instruction initializer.
+
+        Args:
+            duration: Length of time taken by the instruction in terms of dt.
+            *channels: List of pulse channels that this instruction operates on.
+            name: Display name for this instruction.
+        """
+        self._name = name
+
+        self._command = None
+        if not isinstance(duration, int):
+            # TODO: Add deprecation warning once all instructions are migrated
+            self._command = duration
+            if name is None:
+                self._name = duration.name
+            duration = duration.duration
+
+        self._timeslots = TimeslotCollection(*(Timeslot(Interval(0, duration), channel)
+                                               for channel in channels if channel is not None))
+
+    @property
+    def name(self) -> str:
+        """Name of this instruction."""
+        return self._name
+
+    @property
+    def operands(self) -> List[Any]:
+        """Return the operands associated with ``self``."""
+        return [self.duration, self.channels]
```

Comment:
Duration is not necessarily an operand eg., a `FrameChange` will not have an input duration to the instruction (this may even be backend dependent). I view `operands` as what you need to specify to define the instruction.

I think the `channels` property should be specified as an abstract method and implemented by the deriving class. The `Instruction` should accept *operands.

Reply:
operands will be an abstract method